cpp/src/arrow/engine/key_encode.cc

```diff
@@ -0,0 +1,1604 @@
+  ARROW_DCHECK(input.data(buffer_index) != nullptr);
+  ARROW_DCHECK(output.mutable_data(buffer_index) != nullptr);
+
+  util::BitUtil::bytes_to_bits(ctx->instr, static_cast<int>(input.get_length()),
+                               input.data(buffer_index),
+                               output.mutable_data(buffer_index));
+}
+
+bool KeyEncoder::EncoderInteger::IsBoolean(const KeyColumnMetadata& metadata) {
+  return metadata.is_fixed_length && metadata.fixed_length == 0;
+}
+
+bool KeyEncoder::EncoderInteger::UsesTransform(const KeyColumnArray& column) {
+  return IsBoolean(column.get_metadata());
+}
+
+KeyEncoder::KeyColumnArray KeyEncoder::EncoderInteger::ArrayReplace(
+    const KeyColumnArray& column, KeyColumnArray& temp) {
+  if (IsBoolean(column.get_metadata())) {
+    return TransformBoolean::ArrayReplace(column, temp);
+  }
+  return column;
+}
+
+void KeyEncoder::EncoderInteger::PreEncode(const KeyColumnArray& input,
+                                           KeyColumnArray& output,
+                                           KeyEncoderContext* ctx) {
+  if (IsBoolean(input.get_metadata())) {
+    TransformBoolean::PreEncode(input, output, ctx);
+  }
+}
+
+void KeyEncoder::EncoderInteger::PostDecode(const KeyColumnArray& input,
+                                            KeyColumnArray& output,
+                                            KeyEncoderContext* ctx) {
+  if (IsBoolean(output.get_metadata())) {
+    TransformBoolean::PostDecode(input, output, ctx);
+  }
+}
+
+void KeyEncoder::EncoderInteger::Encode(uint32_t* offset_within_row, KeyRowArray& rows,
+                                        const KeyColumnArray& col, KeyEncoderContext* ctx,
+                                        KeyColumnArray& temp) {
+  KeyColumnArray col_prep;
+  if (UsesTransform(col)) {
+    col_prep = ArrayReplace(col, temp);
+    PreEncode(col, col_prep, ctx);
+  } else {
+    col_prep = col;
+  }
+
+  uint32_t num_rows = static_cast<uint32_t>(col.get_length());
+
+  // When we have a single fixed length column we can just do memcpy
+  if (rows.get_metadata().is_fixed_length &&
+      rows.get_metadata().fixed_length == col.get_metadata().fixed_length) {
+    ARROW_DCHECK(*offset_within_row == 0);
+    uint32_t row_size = col.get_metadata().fixed_length;
+    memcpy(rows.mutable_data(1), col.data(1), num_rows * row_size);
+  } else if (rows.get_metadata().is_fixed_length) {
+    uint32_t row_size = rows.get_metadata().fixed_length;
+    uint8_t* row_base = rows.mutable_data(1) + *offset_within_row;
+    const uint8_t* col_base = col_prep.data(1);
+    switch (col_prep.get_metadata().fixed_length) {
+      case 1:
+        for (uint32_t i = 0; i < num_rows; ++i) {
+          row_base[i * row_size] = col_base[i];
+        }
+        break;
+      case 2:
+        for (uint32_t i = 0; i < num_rows; ++i) {
+          *reinterpret_cast<uint16_t*>(row_base + i * row_size) =
+              reinterpret_cast<const uint16_t*>(col_base)[i];
+        }
```

Comment:
There's a lot of unaligned accesses like this one. This is undefined behavior in C++ and it's not supported on all platforms. Could we use SafeLoadAs and SafeStore? If those produce a performance regression, can we optimize them?